### PR TITLE
Add a new PageTree which will be used to display pages list

### PR DESCRIPTION
**Pull request by crazytonyli**
See https://github.com/wordpress-mobile/WordPress-iOS/pull/22048 regarding the performance issue in [the `hierarchySort` function](https://github.com/wordpress-mobile/WordPress-iOS/blob/6d6f1cac01da5c73aadcd890a308ca9b10e9e63d/WordPress/Classes/Extensions/Array%2BPage.swift#L82-L89).

This PR adds a `PageTree` type as a replacement for the `hierarchySort` function. Even though we could replace the `hierarchySort` function right now, the result still isn't good enough to display a site with 1000+ pages. Because currently pages list rebuild the page hierarchy list every time there is a new batch of pages added to the database.

In terms of performance improvements. Here are the result from the unit tests (Updated to reflect some further improvements):
* More than ~120~ 250 times faster when dealing with a flat list where all pages are top-level pages.
* ~25~ 50 times faster when dealing with pages that have a 5 level deep hierarchy.

However, all of the above is test data, we probably still need to improve `PageTree` to make it usable in the real app, which will be my next step.

> [!Note]
> The `PageTree` is currently in the unit test target since it's still a WIP.

## Regression Notes
N/A

### CODE REVIEW

**Comment by crazytonyli** [comment]
Most of added unit tests are focus on the performance. I also want to add unit tests to ensure the result is correct: the pages list is in a correct hierarchical order.

Ideally, we can easy see the page hierarchy in the test code, rather than having to analyze the fake pages data and create a hierarchy in our head. So it'd be nice if we can make this `createFakePages` happen.

```
let pageListHierachy = """
1
|- 2
   |- 3
|- 4
|- 5
   |- 6
7
"""
let pages: [Page] = createFakePages(using: pageListHierachy)
let pageList = algorithm(pages)
assert(pageList, [1,2,3,4,5,6,7])
"""
```

**Comment by crazytonyli** [comment]
Here are a few data points in using the new algorithm in FG:

```
PageTree took 224.321 millisecond to process 1001 pages
PageTree took 504.824 millisecond to process 2001 pages
PageTree took 817.305 millisecond to process 3001 pages
PageTree took 1145.143 millisecond to process 4001 pages
PageTree took 1526.751 millisecond to process 5001 pages
```

File: WordPress/WordPressTest/PagesListTests.swift [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+    }
+}
+
+final class PageTree {
+
+    // A node in a tree, which of course is also a tree itself.
+    private class TreeNode {
+        var page: Page
+        var children = [TreeNode]()
+        var parentNode: TreeNode?
```
Comment (mokagio) [inline WordPress/WordPressTest/PagesListTests.swift]: Of these, the only one that mutates is `parentNode`. Should the other be constants? 

```suggestion
        let page: Page
        let children = [TreeNode]()
        var parentNode: TreeNode?
```

File: WordPress/WordPressTest/PagesListTests.swift [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+
+        try roundtrip("""
+            1
+              2
+                3
+              4
+                5
+              6
+            7
+            """)
```
Comment (mokagio) [inline WordPress/WordPressTest/PagesListTests.swift:139]: This visual language to build the hierarchy is neat and useful.

Nitpick: Using two spaces indentation when the default for the Swift code set in the project is four makes it a bit odd. Have you considered using the same number of spaces?

**Review by mokagio — COMMENTED** [review]
Looking great. Very impressive performance improvements!

> the result still isn't good enough to display a site with 1000+ pages

How many such sites have we got?

As far as I understand, you have more work you want to here, so I guess I should wait to approve?

Also, did you come up with this implementation out of the blue, or are you following some efficient Tree algorithm? If it's the latter, have you considered adding a link for reference?

**Comment by crazytonyli** [comment]
> > the result still isn't good enough to display a site with 1000+ pages
> 
> How many such sites have we got?

I'm not sure. I haven't tried to look to be honest.

> As far as I understand, you have more work you want to here, so I guess I should wait to approve?

This PR is actually rest. The "next steps" I mentioned should be part where we use the new `PageTree` in the UI. Hmmm, actually, I'll make one more change: moving the `PageTree` code out of unit tests to the app target.

> Also, did you come up with this implementation out of the blue, or are you following some efficient Tree algorithm? If it's the latter, have you considered adding a link for reference?

I guess the DFS and BFS is pretty standard algorithm. Other than those, there is not much "efficient algorithm" involved actually.

A tree is typically used to improve searching, where the implementation tries to manipulate the tree structure to maintain a relative shallow height. However, we use a tree structure here for a totally different purpose: displaying the page hierarchy on UI, where we can not change the tree structure, thus will can't have most performance benefits that come with using tree over list.

File: WordPress/WordPressTest/PagesListTests.swift (hunk 2 above)
Comment (crazytonyli) [reply WordPress/WordPressTest/PagesListTests.swift:139]: I feel like using 4 spaces is not as clear, in terms of seeing the hierachy... Here is an example
```
1
    2
3
    4
        5
    6
7
```

File: WordPress/WordPressTest/PagesListTests.swift (hunk 1 above)
Comment (crazytonyli) [reply WordPress/WordPressTest/PagesListTests.swift]: Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/22056/commits/fdd9e7267f1043927fc131edadfe3bc12126bccb. `page` is not mutable, but `children` is.

**Review by mokagio — APPROVED** [review]
Thanks @crazytonyli can't wait to see this in action

**Pull request merged**